### PR TITLE
Fixes #6312: reduces the size of molecule pickles which include atom and bond properties

### LINKED ISSUE #6312: Reduce space overhead of enabling the inclusion of properties when serializing molecules
**Describe the bug**
It seems there may be a significant overhead affecting the size of the Mol binary data if the serialization of additional properties is enabled also when no properties are actually present.

**To Reproduce**
I experienced this issue using the C++ API, but this I think could be a more compact and equivalent reproducible.

```
In [10]: mol = Chem.MolFromSmiles("CCCCC")

In [11]: bin1 = mol.ToBinary()

In [12]: len(bin1)
Out[12]: 98

In [13]: bin2 = mol.ToBinary(
    Chem.PropertyPickleOptions.AtomProps |
    Chem.PropertyPickleOptions.BondProps |
    Chem.PropertyPickleOptions.PrivateProps
)

In [14]: len(bin2)
Out[14]: 146

In [15]: bin3 = mol.ToBinary(Chem.PropertyPickleOptions.AllProps)

In [16]: len(bin3)
Out[16]: 544
```

**Expected behavior**
Ideally, when properties are not assigned to the molecule, passing additional flags enabling their serialization should have no effect on the size of the serialized binary data. This would allow the client code to serialize an etherogeneous collection of molecule and just specify which properties are of interest, without checking if these flags are to be enabled on a per-molecule basis.

**Configuration (please complete the following information):**
 - RDKit version: 2022.09.5

**Additional context**
I don't know if it may be worth considering in relation to #6310 and #6311

### PR DESCRIPTION
**Pull request by greglandrum**
The code changes aren't all that complex: storing key-value pairs is expensive, so we recognize particular keys when present and store then differently.

Using this scheme storing the private properties for atoms and bonds on the 1000 ChEMBL molecules I looked at in the discussion of #6312 looks like this:
```
In [6]: len(b''.join([x.ToBinary() for x in ms]))
Out[6]: 1578735

In [7]: len(b''.join([x.ToBinary(Chem.PropertyPickleOptions.AtomProps|Chem.PropertyPickleOptions.BondProps|Chem.PropertyPickleOption
   ...: s.PrivateProps) for x in ms]))
Out[7]: 2184606
```
Previously just storing either the atom or bond properties was larger than that.

Adding computed properties still takes a lot of space:
```
In [8]: len(b''.join([x.ToBinary(Chem.PropertyPickleOptions.AtomProps|Chem.PropertyPickleOptions.BondProps|Chem.PropertyPickleOption
   ...: s.PrivateProps|Chem.PropertyPickleOptions.ComputedProps) for x in ms]))
Out[8]: 5731323
```
But that's more or less inevitable due to the vector of computed property names which is currently being stored.

We could work around that, but I wonder if it's worth it.

### CODE REVIEW

**Comment by greglandrum** [comment]
@bp-kelley : this includes a modified version of your optimizatoin to only write the `BEGIN_PROPS`/`END_PROPS` tags when they are actually needed.

**Comment by d-b-w** [comment]
I wonder if it would make sense to do this for _all_ atom/bond properties. I mean, the approach here is especially condensed for these specific property type...

But maybe we basically want to store pairs of: `propname, vector<T>` (value for each atom) or `propname, vector<pair<int, T>>`, (index of atom with the property + its value). It sort of depends on how dense properties tend to be - like do people set a property on _all_ atoms, or just a few? I think that it is true that people rarely set a property on just one atom.

**Comment by greglandrum** [comment]
> I wonder if it would make sense to do this for _all_ atom/bond properties. I mean, the approach here is especially condensed for these specific property type...
> 
> But maybe we basically want to store pairs of: `propname, vector<T>` (value for each atom) or `propname, vector<pair<int, T>>`, (index of atom with the property + its value). It sort of depends on how dense properties tend to be - like do people set a property on _all_ atoms, or just a few? I think that it is true that people rarely set a property on just one atom.

I think it's definitely worth thinking about alternate ways of storing these things (both in the pickles and on the molecules themselves). What we're currently doing is extremely flexible but very memory intensive.

Until we've had a chance to do that, I still think it makes sense to merge this change

**Comment by greglandrum** [comment]
@bp-kelley I think this ought to be merged now that we're on the other side of the 2023.03 release.
Can you please take a look?

File: Code/GraphMol/MolPickler.cpp [hunk 1]
```diff
@@ -106,6 +107,163 @@ void write_sstream_to_stream(std::ostream &outStream,
...
+  PropTracker() {
+    for (const auto &pr : explicitBondProps) {
+      ignoreBondProps.insert(pr.first);
+    }
+  };
+};
+
+bool pickleAtomProperties(std::ostream &ss, const RDProps &props,
+                          unsigned int pickleFlags) {
+  const static PropTracker aprops;
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.cpp:199]: Yep this is weird!  You could have also done something like

```
bool pickleAtomProperties(std::ostream &ss, const RDProps &props,
                           unsigned int pickleFlags, const PropTracker &aProps = PropTracker())
```

**Review by bp-kelley — APPROVED** [review]

**Pull request merged**